### ISSUE TITLE
decoding objects with optional keys

### ISSUE THREAD

**Issue by CiprianMate** [ooesili/type-safe-json-decoder#8]
Hi!

Is it possible to decode an object witch has optional keys.

For example, i have an object array with the fallowing object structure:
{
  id: string;
  label: string;
  inputType: textarea;
  readonly: boolean; - optional key
}

i want to be able to decode the same object with or without the readonly key.

as a workaround I've made 2 decoders, one with and one without the readonly key, and i'm using oneOf (decoder list)

Is there any other way to do this? the way i did is unmanageable if you have multiple optional key.

Thanks!

**Comment by mrmurphy** [comment]
@CiprianMate, this is a bit messy, but you should be able to achieve what you're going for with something like this:

```
andThen(
    oneOf(at(['readOnly'], boolean()), succeed(false))
    /* ☝️ if there is no 'readOnly' key, the value 'false' will be used instead */
    , readOnly => {
        object(
            ["id", string()],
            ["label", string()],
            ["inputType", oneOf(equal("textarea"), equal("text")/*, whatever other input types you want */)],
            (id, label, inputType) => ({
                id,
                label,
                inputType,
                readOnly
            })
        )
    }
)
```

And you could probably write a helper function to help you do that multiple times without too much horrible boilerplate.

@ooesili It seems to me that it might be reasonable for the `object` pass a null to a decoder when the field doesn't exist, though that's a departure from the Elm way if I remember correctly, and it's nice to know which fields are missing.

Here's an idea that I haven't thought out well: What if giving a key a question mark as a suffix made the `object` decoder pass a null for the value of that key if the key doesn't exist, instead of reporting that the key isn't there? Something like this:
```
object(['foo?', string()], (foo: string | null) => {foo})
```

I don't love that it becomes "stringly typed", but it might be a quick and simple way to address this problem.

**Comment by ooesili** [comment]
I created #10 to try and solve this. It comes pretty close to what Elm does, but instead of `Maybe a` it's `a | undefined`. I chose undefined instead of null since `field?` in an interface is the same thing as `field | undefined`.

**Comment by ooesili** [comment]
Just merged #10. Closing.
